File: daily/Sliding Window/904.Fruit-Into-Baskets.go
package daily

func totalFruit(fruits []int) int {
	type_fruit := make(map[int]int)
	res, left := 0, 0
	for right, _ := range fruits {
		type_fruit[fruits[right]]++

		if len(type_fruit) > 2 {
			type_fruit[fruits[left]]--

			if type_fruit[fruits[left]] == 0 {
				delete(type_fruit, fruits[left])
			}
			left++
		}

		res = max(res, right-left+1)
	}
	return res
}

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

File: daily/Sliding Window/904.Fruit-Into-Baskets.py
from typing import List


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        count = {}
        left = 0
        best = 0

        for right, f in enumerate(fruits):
            count[f] = count.get(f, 0) + 1

            while len(count) > 2:
                lf = fruits[left]
                count[lf] -= 1
                if count[lf] == 0:
                    del count[lf]
                left += 1

            best = max(best, right - left + 1)

        return best

